src/lib.rs

```diff
@@ -182,6 +182,49 @@ impl iter::FromIterator<char> for SmolStr {
     }
 }
 
+fn build_from_str_iter<T>(mut iter: impl Iterator<Item=T>) -> SmolStr
+where
+    T: AsRef<str>,
+    std::string::String: std::iter::Extend<T>,
+{
+    use std::io::prelude::*;
+
+    let mut len = 0;
+    let mut buf = [0u8; INLINE_CAP];
+    while let Some(slice) = iter.next() {
+        let slice = slice.as_ref();
+        let size = slice.len();
+        if size + len > INLINE_CAP {
+            let mut heap = String::with_capacity(size + len);
+            heap.push_str(std::str::from_utf8(&buf[..len]).unwrap());
+            heap.push_str(&slice);
+            heap.extend(iter);
+            return SmolStr(Repr::Heap(heap.into_boxed_str().into()));
+        }
+        (&mut buf[len..]).write_all(slice.as_bytes()).unwrap();
+        len += size;
+    }
+    SmolStr(Repr::Inline { len: len as u8, buf })
+}
```

Comment:
Hm....

I feel this is subtly incorrect, in that before this change, `Repr` was canonical: equal strings had equal `Repr`s (and, in fact, `PartialEq` could be optimized to compare reprs without going via `as_str()`).

Now this is not the case, because `Repr::Substring` is never created here.

Hm, **maybe** this problem was introduce before, when `FromIterator<char>` was added? 

I am not sure if it worth to go out of our way to preserve canonicity of Repr, but, without thinking to much about it, I would prefer to fix `FromIterator` impls to correctly deal with substring case